File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    let dummy=new ListNode(0,head);
    let curr=dummy;
    let second=dummy;
    for(let i=0;i<n+1;i++){
        curr=curr.next;
    }
    while(curr!=null){
        curr=curr.next;
        second=second.next;
    }
    second.next=second.next.next;
    return dummy.next;
};

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        // Pehle length nikal lo
        int length = 0;
        ListNode curr = head;
        while (curr != null) {
            length++;
            curr = curr.next;
        }

        // Agar first node delete karni ho
        if (n == length) {
            return head.next;
        }

        // Traverse karke us node ke pehle tak jao
        curr = head;
        for (int i = 1; i < length - n; i++) {
            curr = curr.next;
        }

        // Ab deleteNode jaisa logic lagao
        curr.next = curr.next.next;

        return head;
    }
}
